vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp

```diff
@@ -0,0 +1,312 @@
+    long found = 0;
+    rnd.srand48(++_read_seed);
+    uint32_t i;
+    for (i = 0; i < cnt && _stop_read.load() == 0; ++i) {
+        auto guard = _generation_handler.takeGuard();
+        uint32_t key = rnd.lrand48() % (_keyLimit + 1);
+        MyCompare comp(_store, key);
+        auto hash_map = _hash_map.load(std::memory_order_acquire);
+        auto result = hash_map->find(comp, EntryRef());
+        if (result != nullptr) {
+            auto ref = result->first.load_relaxed();
+            auto &wrapped_entry = _allocator.get_wrapped(ref);
+            EXPECT_EQ(key, wrapped_entry.value());
+            ++found;
+        }
+    }
+    _done_read_work += i;
+    _found_count += found;
+    LOG(info, "done %u read work", i);
+}
+
+
+void
+DataStoreFixedSizeHashTest::read_work()
+{
+    read_work(std::numeric_limits<uint32_t>::max());
+}
+
+
+void
+DataStoreFixedSizeHashTest::write_work(uint32_t cnt)
+{
+    vespalib::Rand48 &rnd(_rnd);
+    for (uint32_t i = 0; i < cnt; ++i) {
+        uint32_t key = rnd.lrand48() % _keyLimit;
+        if ((rnd.lrand48() & 1) == 0) {
+            insert(key);
+        } else {
+            remove(key);
+        }
+        commit();
+    }
+    _done_write_work += cnt;
+    _stop_read = 1;
+    LOG(info, "done %u write work", cnt);
+}
+
+
+TEST_F(DataStoreFixedSizeHashTest, smoke_test)
+{
+    EXPECT_EQ(0, size());
+    insert(1);
+    EXPECT_EQ(1, size());
+    remove(2);
+    EXPECT_EQ(1, size());
+    insert(1);
+    EXPECT_EQ(1, size());
+    insert(5);
+    EXPECT_EQ(2, size());
+    insert(4);
+    EXPECT_EQ(3, size());
+    remove(3);
+    EXPECT_EQ(3, size());
+    remove(5);
+    EXPECT_EQ(2, size());
+    commit();
+    MyCompare comp3(_store, 3);
+    auto hash_map = _hash_map.load(std::memory_order_acquire);
+    auto result3 = hash_map->find(comp3, EntryRef());
+    EXPECT_TRUE(result3 == nullptr);
+    MyCompare comp4(_store, 4);
+    auto result4 = hash_map->find(comp4, EntryRef());
+    EXPECT_TRUE(result4 != nullptr);
+    auto ref4 = result4->first.load_relaxed();
+    auto& wrapped_entry4 = _allocator.get_wrapped(ref4);
+    EXPECT_EQ(4, wrapped_entry4.value());
+}
+
+TEST_F(DataStoreFixedSizeHashTest, lookups_works_after_insert_and_remove)
+{
+    _modulo_limit = 1; // Force single hash chain
+    vespalib::hash_map<uint32_t, bool> expected;
+    vespalib::Rand48 &rnd(_rnd);
+    for (uint32_t i = 0; i < 40; ++i) {
+        uint32_t key = rnd.lrand48() % 10;
+        if ((rnd.lrand48() & 1) == 0) {
+            insert(key);
+            expected[key] = true;
+        } else {
+            remove(key);
+            expected[key] = false;
+        }
+        commit();
+    }
+    auto hash_map = _hash_map.load(std::memory_order_acquire);
+    for (auto &kv : expected) {
+        MyCompare comp(_store, kv.first);
+        EXPECT_EQ(kv.second, hash_map->find(comp, EntryRef()) != nullptr);
+    }
+}
+
+TEST_F(DataStoreFixedSizeHashTest, single_threaded_reader_without_updates)
```

Comment:
This is mainly covered in simple_hash_map_test.cpp, consider removing.

Reply:
Removed.